server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -478,6 +479,19 @@ private void executeRequest(
         Rewriteable.rewriteAndFetch(original, searchService.getRewriteContext(timeProvider::absoluteStartMillis), rewriteListener);
     }
 
+    static void adjustSearchType(SearchRequest searchRequest) {
+        // if there's a kNN search, always use DFS_QUERY_THEN_FETCH
```

Comment:
This means that if `knn` section present, dfs stage will always be applied to `query` part  even if we don't need it. Do you think we should invoke dfs stage for query only when explicitly requested?

Reply:
Right, in the docs I planned to present this as "if `knn` is provided, then the `dfs_query_then_fetch` type is always used". I thought this was easiest to understand and kept the implementation simple:
* Knowing that `dfs_query_then_fetch` is used helps explain why things like `ccs_minimize_roundtrips` will not work. kNN has all the same limitations as if `dfs_query_then_fetch` was set.
* It feels weird to run the DFS search phase but not compute DFS?

In my understanding, the main overhead of DFS is from the extra roundtrip, not gathering the stats, so it won't be too expensive. Interested to know your opinion!

Reply:
> "if knn is provided, then the dfs_query_then_fetch type is always used"

This sounds good to me. 

> It feels weird to run the DFS search phase but not compute DFS?

What do you mean by "not compute DFS" ? It looks to me if `query` and `knn` search are present, we always compute and use DFS stats information? right?

Reply:
Thanks for clarifications, Julie!